crates/typed-store/src/raw_key_db_map.rs

```diff
@@ -0,0 +1,498 @@
+                Ok(Some(value))
+            }
+            None => Ok(None),
+        }
+    }
+
+    /// Gets a value within a transaction.
+    pub fn get_cf_with_txn(
+        &self,
+        txn: &rocksdb::Transaction<'_, OptimisticTransactionDB>,
+        key: &K,
+    ) -> Result<Option<V>, TypedStoreError> {
+        let key_bytes = key.serialize()?;
+        let cf = self.inner.cf()?;
+
+        let value_bytes = txn
+            .get_cf(&cf, &key_bytes)
+            .map_err(|e| TypedStoreError::RocksDBError(e.to_string()))?;
+
+        match value_bytes {
+            Some(bytes) => {
+                let value = bcs::from_bytes(&bytes)
+                    .map_err(|e| TypedStoreError::SerializationError(e.to_string()))?;
+                Ok(Some(value))
+            }
+            None => Ok(None),
+        }
+    }
+
+    /// Reads a range of key-value pairs within a transaction using efficient range bounds.
+    /// Returns key-value pairs in the range [begin, end).
+    ///
+    /// This is more efficient than prefix-based iteration for range queries.
+    pub fn read_range(
+        &self,
+        txn: &rocksdb::Transaction<'_, OptimisticTransactionDB>,
+        begin: Vec<u8>,
+        end: Vec<u8>,
+        row_limit: usize,
+    ) -> Result<Vec<(K, V)>, TypedStoreError> {
+        let cf = self.inner.cf()?;
+
+        let mut read_opts = rocksdb::ReadOptions::default();
+        read_opts.set_iterate_lower_bound(begin);
+        read_opts.set_iterate_upper_bound(end);
+
+        let iter = txn.iterator_cf_opt(&cf, read_opts, rocksdb::IteratorMode::Start);
+
+        let mut results = Vec::new();
+
+        for item in iter.take(row_limit) {
+            match item {
+                Ok((key_bytes, value_bytes)) => {
+                    match (|| -> Result<(K, V), TypedStoreError> {
+                        let key = K::deserialize(&key_bytes)?;
+                        let value = bcs::from_bytes(&value_bytes)
+                            .map_err(|e| TypedStoreError::SerializationError(e.to_string()))?;
+                        Ok((key, value))
+                    })() {
+                        Ok(kv_pair) => results.push(kv_pair),
+                        Err(e) => {
+                            if results.is_empty() {
+                                return Err(e);
+                            } else {
+                                break; // Return partial results
+                            }
+                        }
+                    }
+                }
+                Err(e) => {
+                    if results.is_empty() {
+                        return Err(TypedStoreError::RocksDBError(e.to_string()));
+                    } else {
+                        break; // Return partial results
+                    }
+                }
+            }
+
+            if results.len() >= row_limit {
+                break;
+            }
+        }
+
+        Ok(results)
```

Comment:
suggestion: can we return an iterator here so that we don't need to hold all the results in memory before returning?

Reply:
Yeah, we can re-evaluate if this ended up concuming a lot memory.